File: src/main/java/io/gittul/domain/post/dto/PostDetailResponse.java
package io.gittul.domain.post.dto;

import io.gittul.domain.comment.dto.CommentResponse;
import io.gittul.domain.post.entity.Post;
import io.gittul.domain.tag.entity.Tag;
import io.gittul.domain.user.dto.UserProfileResponse;
import io.gittul.domain.user.entity.User;

import java.math.BigInteger;
import java.util.List;

public record PostDetailResponse(
        BigInteger id,
        String title,
        String content,
        String image,
        String createdAt,
        String updatedAt,
        UserProfileResponse writer,
        boolean isFollowed,
        boolean isLiked,
        boolean isBookmarked,
        int likeCount,
        int bookmarkCount,
        int viewCount,
        List<String> tags,
        List<CommentResponse> comments
) {
    public static PostDetailResponse of(Post post, User requestingUser) {
        return new PostDetailResponse(
                BigInteger.valueOf(post.getPostId()),
                post.getTitle(),
                post.getContent(),
                post.getImageUrl(),
                post.getCreatedAt().toString(),
                post.getUpdatedAt().toString(),
                UserProfileResponse.of(post.getUser()),
                requestingUser.getDetails().isFollowing(post.getUser()),
                post.isLikedBy(requestingUser),
                post.isBookmarkedBy(requestingUser),
                post.getLikeCount(),
                post.getBookmarkCount(),
                post.getViewCount(),
                post.getTags().stream().map(Tag::getTagName).toList(),
                post.getComments().stream().map(comment -> CommentResponse.ofAndTo(comment, post.getUser())).toList()
        );
    }
}

File: app/src/main/kotlin/io/gittul/app/domain/post/dto/PostDetailResponse.kt
package io.gittul.app.domain.post.dto

import io.gittul.app.domain.comment.dto.CommentResponse
import io.gittul.app.domain.user.dto.UserProfileResponse
import io.gittul.core.domain.post.entity.Post
import io.gittul.core.domain.user.entity.User
import java.math.BigInteger

data class PostDetailResponse(
    val id: BigInteger,
    val title: String,
    val content: String,
    val image: String?,
    val createdAt: String,
    val updatedAt: String,
    val writer: UserProfileResponse?,
    val isFollowed: Boolean,
    val isLiked: Boolean,
    val isBookmarked: Boolean,
    val likeCount: Int,
    val bookmarkCount: Int,
    val viewCount: Int,
    val tags: List<String>,
    val comments: List<CommentResponse>
) {
    companion object {
        fun of(post: Post, requestingUser: User): PostDetailResponse {
            return PostDetailResponse(
                id = BigInteger.valueOf(post.postId),
                title = post.title,
                content = post.content,
                image = post.imageUrl,
                createdAt = post.createdAt.toString(),
                updatedAt = post.updatedAt.toString(),
                writer = UserProfileResponse.of(post.user),
                isFollowed = requestingUser.isFollowing(post.user),
                isLiked = post.isLikedBy(requestingUser),
                isBookmarked = post.isBookmarkedBy(requestingUser),
                likeCount = post.likeCount,
                bookmarkCount = post.bookmarkCount,
                viewCount = post.viewCount,
                tags = post.tags.map { it.tagName },
                comments = post.comments.mapNotNull { CommentResponse.ofAndTo(it, post.user) }
            )
        }
    }
}
